Ullaakut/astronomer
# Generate signed reports

## Goal of this PR

This PR is the most important yet. It makes Astronomer sign and send its trust reports to [Astrolab](https://github.com/ullaakut/astrolab), a server in charge of collecting Astronomer reports and serving GitHub badges.

In order to achieve this, many changes were necessary:

* Most of the code is now separated into public packages
	* `github.com/ullaakut/astronomer/pkg/context`
		* Only contains the astronomer context, used by all other packages to contain data on the current scan
	* `github.com/ullaakut/astronomer/pkg/gql`
		* Exposes the `FetchStargazers` and `FetchContributions` public methods
		* Is in charge to fetching data from the GitHub GraphQL API
		* Handles caching of this data
	* `github.com/ullaakut/astronomer/pkg/trust`
		* Is in charge of computing and rendering trust through the `Compute` and `Render` public methods
	* `github.com/ullaakut/astronomer/pkg/signature`
		* Is used to sign and send trust reports to the Astrolab server
* It is no longer possible for people to build the Astronomer binary by themselves, they must use the released binaries or the public Docker images from DockerHub. This is due to the signing process.
* In order to avoid people using the `--all` to get better trust reports, despite their early users being bot accounts, any scan of a repository with more than 200 users will compare the first 200 users to the rest in a comparative report, and will select the worst trust factors of both reports.
* Astronomer now outputs the URL of the generated GitHub badge from the current scan

Fixes #14 
Fixes #22 

## How to test it

* `go test ./...`
* Run astronomer on a repository with less than 200 stars
* Run astronomer on a repository with more than 200 stars
* Verify that github badges are generated properly

## Examples

<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fastronomer.ullaakut.eu%2Fshields%3Fowner%3Dcontainous%26name%3Dtraefik" />
<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fastronomer.ullaakut.eu%2Fshields%3Fowner%3Doperator996%26name%3Dyaocl" />
<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fastronomer.ullaakut.eu%2Fshields%3Fowner%3Denvoyproxy%26name%3Denvoy" />
<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fastronomer.ullaakut.eu%2Fshields%3Fowner%3Dullaakut%26name%3Dcameradar" />
<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fastronomer.ullaakut.eu%2Fshields%3Fowner%3Dkataras%26name%3Diris" />
